RFD: auto-generate the `tctl` resource reference

ptgott opened the pull request.

ptgott committed:
RFD: auto-generate the `tctl` resource reference

codingllama commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,554 @@
+
+We don't have a formal naming convention for the root types of each resource,
+and may not want to list *all* resources (e.g., there may be internal resources
+to keep hidden), so it would make sense to list resources explicitly. The CRD
```
+1 for the explicit mapping. It shouldn't be too much of a burden to keep.

codingllama commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,554 @@
+However, we can't guarantee that every configuration resource available to
+`tctl` will correspond to a protobuf message definition. Since we generate Go
+source files based on our `.proto` files, and Teleport needs to unmarshal all
+configuration resources into structs from YAML, it's a safe bet that all
+configuration resources will have a corresponding Go struct.
```
+1 for parsing Go sources too. It's pretty unlikely a resource won't have a Go representation, plus it doesn't tie us to necessarily to protos (or a particular proto generator).

codingllama commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,554 @@
+`singlechecker`
+(https://pkg.go.dev/golang.org/x/tools/go/analysis/singlechecker) packages
+within `golang.org/x/tools/go/analysis`, which run custom static analysis
+functions, `analysis.Analyzer`s. However, these static analysis tools are
+intended for linting, not generating reference documentation. 
```
> However, these static analysis tools are intended for linting, not generating reference documentation

Design intention is key. It might work for a while, but it seems poor form to depend on it.

A program as simpler as we can make it, without unnecessary attachments, seems better long-term.

codingllama commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,554 @@
+|`Resource.SourcePath`|The `file` property of the underlying `sourceDecl`.|
+|`Resource.Description`|`GenDecl.Doc`|
+|`Resource.SectionName`|`TypeSpec.Name.Name`, with camelcase converted to separate words.|
+|`Fields[*].Name`|Process the `Field.Tag` in the `FieldList`, extracting the value of the `json` tag.|
```
Do we expect json tags to always exist?

ptgott replied:
Ah, that's a good point. Should the generator check whether the struct type has an `UnmarshalYAML` or `UnmarshalJSON` method before processing the struct declaration AST? 

If there is an `UnmarshalYAML` or `UnmarshalJSON` method, the generator would treat the struct as a custom type, since it won't be able to tell how struct field names translate into YAML field names. If there are no such methods, the generator would use the original struct field name.

codingllama replied:
I'm more interested in what our "requirements" are - it's alright if we want all structs to have JSON tags, as long as we say so here.

Special handling of custom unmarshal methods seems like a good idea, or at the very least some warnings from the generator.

codingllama commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,554 @@
+
+#### Custom fields
+
+Entries in the reference for custom fields, e.g., `Labels`, don't lend
+themselves to automatic generation because they implement custom unmarshalers,
+and we can't rely on `json` struct tags to indicate the types of these fields.
```
Could we make an educated guess and provide a way to override it? So simple numeric/string fields are taken from the Go type, as well as references to other structs?

Edit: I think that's what you explained below.

ptgott replied:
I'll add my response to this in my response to [this comment](https://github.com/gravitational/teleport/pull/25823#discussion_r1191510099).

codingllama commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,554 @@
+authors: Paul Gottschling (paul.gottschling@goteleport.com)
+state: draft
+---
+
```
Could you a required approvers section?

https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md?plain=1#L106

codingllama reviewed (commented):
Thanks for writing the RFD, Paul. Overall seems reasonable, although I confess I didn't spend too much time considering the implementation, as it seems alright for us to refine it as we go.

Would you mind following the RFD numbering/branching process described at https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md?plain=1#L58-L78? I think we can keep this PR, since it already has some history, but maybe take the numbered branch too so it's clear it exists (and point it back to here?)

codingllama commented:
Also, feel free to move out of draft, or alternatively let us know what's keeping you.

ptgott commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,554 @@
+modify it for our purposes, the reference generator does not require the
+general-purpose static analysis capabilities of the `analysis` package.
+
+### Converting structs to `Resource`s
```
For now, I think the only requirement is that a GoDoc exists. I've suggested exiting with an error if there's no GoDoc above a field so we can add one. 

The only special rule is that, if a field name begins with the name of the field, remove the field name (and "is" if it exists) and capitalize the next first letter. This is to prevent field names that aren't user facing from making it into the docs, and also removes repetition with the first column of the table.

This GoDoc:

```go
// MyField is a field that specifies something.
```

...would become:

```
A field that specifies something
```

Otherwise, I think it's okay to move each field's GoDoc into the table without alteration.

codingllama replied:
> if a field name begins with the name of the field, remove the field name

The caveat I see here is that sometimes the field name is part of the sentence in a way that could have this break, but I suppose we can be careful and edit the types we care about.

ptgott replied:
> The caveat I see here is that sometimes the field name is part of the sentence in a way that could have this break, but I suppose we can be careful and edit the types we care about.

We could also simply replace all mentions of the original field name in the GoDoc for a field with mentions of the one included in the docs, i.e., the struct tag. I'll include that in the RFD.

ptgott commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,554 @@
+themselves to automatic generation because they implement custom unmarshalers,
+and we can't rely on `json` struct tags to indicate the types of these fields.
+
+For simplicity, we can describe custom fields by hardcoding their descriptions
+and example YAML values. To do so, we can include an `examples` directory within
```
In this case, would it make sense to add an example YAML document for a custom type within the type's GoDoc?

E.g., for [`Labels`](https://github.com/gravitational/teleport/blob/c91da46765953688d218dcba4a4ce7acf606639f/api/types/role.go#L1137-L1140):

```go
// Labels is a wrapper around map
// that can marshal and unmarshal itself
// from scalar and list values
// Example YAML:
// 'env': 'test'
// '*': '*'
// 'region': ['us-west-1', -eu-central-1]
// 'reg': ^us-west-1|eu-central-1$'
// ---
type Labels map[string]utils.Strings
```

The generator could then exit with an error if a GoDoc for a custom type does not include an example YAML document, which it interprets as everything between `Example YAML` and the following `---`. If it can't parse the example YAML document as YAML, it also exits with an error.

codingllama replied:
That sounds fine to me, +1.

ptgott committed:
Partial response to codingllama feedback

codingllama commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -552,3 +560,9 @@ table above the example YAML) to see an example YAML document for this type.
+
+## Test plan
+
+We can add an item to the "Documentation" section of the test plan ensuring that
```
Assuming this is not already a part of our release procedure, +1.

Better yet if we can make it part of the release procedure, I think.

codingllama commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,568 @@
+
+## Required Approvers
+
+- Engineering: @codingllama
```
Add a 2nd approver for another pair of eyes? Maybe @zmb3, since he's already reviewed.

codingllama reviewed (approved):
Thanks for putting this together, Paul.

No overriding concerns on my part.

codingllama commented:
> Would you mind following the RFD numbering/branching process described at https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md?plain=1#L58-L78? I think we can keep this PR, since it already has some history, but maybe take the numbered branch too so it's clear it exists (and point it back to here?)

Friendly ping on this? @ptgott

ptgott committed:
Partially respond to feedback

- Be more specific about expectations around struct tags
- Propose replacing struct field names in field descriptions with the
  user-facing field name
- Use comments instead of separate files for example YAML
- Be more explicit about the scope of the RFD (it applies to resources
  applied via `tctl`)
- Use a Go map for the generator configuration
- Indicate that this Make target will depend on `make grpc`.
- Add a second Eng reviewer
- Propose making running the generator part of our release procedure

ptgott commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,568 @@
+
+## What
+
+Automatically generate the "Dynamic resources" section of the configuration
```
That's right, the scope of this RFD is for resources you can apply via `tctl`. The terminology mistake is also based on my own misconceptions, though I'm not sure the current docs are clear on this either.

It looks like we have three different uses of "resource" in the context of Teleport:
- Dynamic configuration resources (i.e., alternatives to static configuration files, as you mention)
- Teleport resources, e.g., `role`s and local `user`s
- Infrastructure resources you can enroll with Teleport, either via Teleport resources or configuration files

I think these three usages are okay as long as we make sure we're precise about them. I know this is off-topic, but @zmb3 should we mention these distinctions in the Core Concepts page?

ptgott commented on rfd/xxxx-config-resource-docs.md:
```diff
@@ -0,0 +1,568 @@
+
+We don't have a formal naming convention for the root types of each resource,
+and may not want to list *all* resources (e.g., there may be internal resources
+to keep hidden), so it would make sense to list resources explicitly. The CRD
```
I suggested manually listing the resources because that's how we do it with the [Kubernetes operator](https://github.com/gravitational/teleport/blob/09adc3ea5cddac5c1fea479ddbfc0075c864677f/integrations/operator/crdgen/main.go#L97-L113). Since we need to develop controllers for the operator, this makes sense in that context, but you're right that it's not great for `tctl`.

I like the idea of including all resources with an embedded `Metadata` field.

ptgott replied:
@codingllama @zmb3 Come to think of it, are there any structs with a `github.com/gravitational/teleport/api/types.Metadata` field that we would want to exclude? Would it be possible to eliminate the configuration entirely?

codingllama replied:
Not sure, but that should be easy to search for. If there aren't any, +1 to automatically finding resources.

ptgott committed:
Propose including all Teleport resources

And excluding user-configured resources
